.github/actions/update-extension-provider-images/update-images.py

```diff
@@ -0,0 +1,626 @@
+    - For new minor/major versions, it adds a new image entry.
+    - It ensures the highest minor version has a `targetVersion` of ">= X.Y".
+    - For images without a `targetVersion`, it keeps only the single latest tag.
+5.  It generates a `release-notes.md` file containing direct links to the
+    release pages for all new and intermediate tags.
+
+This script is not intended to be run manually by developers. It should be
+invoked via the `gardener/cc-utils/.github/workflows/update-extension-provider-images.yaml`
+reusable workflow.
+
+Manual Intervention:
+This script handles version *updates*. Manually editing `images.yaml` is still
+required to remove entire image groups for deprecated Kubernetes versions that
+are no longer supported.
+Image Updater Script.
+"""
+
+import argparse
+import yaml
+import re
+import sys
+import semver
+import time
+from typing import Dict, List, Tuple, Optional, TypedDict
+from collections import defaultdict
+from oci import client as oci_client
+from oci.model import OciImageNotFoundException
+from version import parse_to_semver, is_final
+
+
+# --- Type Definitions ---
+class Update(TypedDict):
+    """A structured representation of a single image update."""
+
+    image_name: str
+    old_tag: Optional[str]
+    new_tag: str
+    update_type: str  # 'patch', 'minor', or 'singleton'
+
+
+# --- Helper Functions ---
+def get_tags_from_registry(oci_client: oci_client.Client, repository: str) -> List[str]:
+    """
+    Use the oci.client to retrieve a list of tags for a repository.
+
+    Args:
+        oci_client (oci_client.Client): An initialized OCI client.
+        repository (str): The image repository to query.
+
+    Returns:
+        List[str]: A list of tag strings, e.g., ['v1.2.3', 'v1.2.4'].
+                   Returns an empty list on error.
+    """
+    try:
+        return oci_client.tags(image_reference=repository)
+    except Exception as e:
+        print(f"Error retrieving tags for {repository}: {e}", file=sys.stderr)
+        return []
+
+
+def image_exists(oci_client: oci_client.Client, repository: str, tag: str) -> bool:
+    """
+    Check if a specific image tag exists in the registry by trying to fetch its manifest.
+
+    Args:
+        oci_client (oci_client.Client): An initialized OCI client.
+        repository (str): The image repository.
+        tag (str): The image tag.
+
+    Returns:
+        bool: True if the image exists, False otherwise.
+    """
+    image_ref = f"{repository}:{tag}"
+    retries = 3
+    delay = 2  # seconds
+
+    for i in range(retries):
+        try:
+            oci_client.manifest(image_reference=image_ref, accept="*/*")
+            return True
+        except OciImageNotFoundException:
+            print(f"  ✗ Manifest for {image_ref} not found.", file=sys.stderr)
+            return False
+        except Exception as e:
+            # Check if the error is a rate-limiting error
+            if "429" in str(e) and "Too Many Requests" in str(e):
```

Comment:
will you run queries against gardener-project's GAR exclusively, or will you also target other OCI-Registries, such as docker-hub, quay.io, and the like?

In either way, preferably use the more specific exception raised from `requests` (requests.exceptions.HTTPError, I think?). This exception contains an int w/ status-code, so you do not have to parse / guess. Also, it should be sufficient to only check for status-code.

Consider running `oci-auth`-action prior to calling this script, such as to allow the oci-client to run authenticated requests (this might be helpful, especially if issueing requests against docker-hub - readonly-credentials for DH are implicitly included from said action..)

Reply:
I remove the `image_exists` fctn for now, see [this comment](https://github.com/gardener/cc-utils/pull/1367#discussion_r2362759276)